src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs

```diff
@@ -2538,6 +2578,53 @@ private void ClassifyUpdate(SyntaxNode oldNode, SyntaxNode newNode)
                 }
             }
 
+            private void ClassifyUpdate(GlobalStatementSyntax? oldNode, GlobalStatementSyntax? newNode)
+            {
+                if (newNode is not null)
+                {
+                    ClassifyDeclarationBodyRudeUpdates(newNode.Statement);
+                }
+
+                var otherCompilationUnit = newNode is null ? _match?.NewRoot as CompilationUnitSyntax : _match?.OldRoot as CompilationUnitSyntax;
+                var nodeToAnalyze = newNode ?? oldNode;
+
+                Debug.Assert(otherCompilationUnit is not null);
+                Debug.Assert(nodeToAnalyze is not null);
+
+                // Any change to a global statement could mean the synthesized main method changes return type
+                // from void to int, Task or Task<int> (or vice versa), so we might need to issue a TypeUpdate rude edit.
+                foreach (var node in nodeToAnalyze.DescendantNodesAndSelf(LambdaUtilities.IsNotLambda))
+                {
+                    switch (node.Kind())
+                    {
+                        case SyntaxKind.AwaitExpression:
+                            // If we found an await expression, there must be one in the other root
+                            // or it would mean a change of return type from non-Task to Task (or vice versa)
+                            if (!otherCompilationUnit.DescendantNodesAndSelf(IsNotLambdaOrDeclaration).OfType<AwaitExpressionSyntax>().Any())
+                            {
+                                ReportError(RudeEditKind.TypeUpdate, newNode, newNode ?? oldNode);
+                                return;
+                            }
+                            break;
+                        case SyntaxKind.ReturnStatement:
+                            // If we found a return statement with a null expression, there must be one in the other root, or
+                            // if we found a return statement with a non-null expression, there must be one in the other root
+                            // or it would mean a change of return type from void or Task to int or Task<int> (or vice versa)
+                            var returnExpression = (ReturnStatementSyntax)node;
+                            if (!otherCompilationUnit.DescendantNodesAndSelf(IsNotLambdaOrDeclaration).OfType<ReturnStatementSyntax>().Where(r => (r.Expression == null) == (returnExpression.Expression == null)).Any())
+                            {
+                                ReportError(RudeEditKind.TypeUpdate, newNode, newNode ?? oldNode);
+                                return;
+                            }
+                            break;
+                    }
+                }
+
```

Comment:
Wouldn't it be simpler to check for return type change in semantic analysis?

Reply:
Yes. I originally did it here just so I wouldn't have to do the semantic check for inserts, deletes and updates, but in light of other news this clearly should move :)